gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -504,6 +508,162 @@ def _get_variant_calls(self, record, formats):
+  class - we could only use a single pipe, but it will divert the parsers.
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               file_pattern=None,  # type: str
+               representative_header_lines=None,  # type:  List[str]
+               splittable_bgzf=False,  # type: bool
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PySamParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    self._header_lines = []
+    self._next_line_to_process = None
+    self._current_line = None
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    self._to_child = None
+
+  def _init_with_header(self, header_lines):
+    # The first header line must be similar to '##fileformat=VCFv.*'.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    p_read, c_write = os.pipe()
+    c_read, p_write = os.pipe()
+
+    processid = os.fork()
+    if processid:
+      os.close(c_read)
+      os.close(c_write)
+      into_pysam = os.fdopen(p_read)
+      self._to_child = os.fdopen(p_write, 'w')
+      self._vcf_reader = libcbcf.VariantFile(into_pysam, 'r')
+    else:
+      os.close(p_read)
+      os.close(p_write)
+      from_parent = os.fdopen(c_read)
+      to_pysam = os.fdopen(c_write, 'w')
+      # Feed the data with a break line, since we read data line by line - if
+      # we would use the conventional ``read()`` method, both pipes would lock,
+      # while waiting for an EOF signal.
+      for header in header_lines:
+        to_pysam.write(header.strip() + '\n')
+      to_pysam.flush()
+
+      while 1:
+        text_line = from_parent.readline()
+        if not text_line:
+          break
+        to_pysam.write(text_line)
+        to_pysam.flush()
+
+      from_parent.close()
+      to_pysam.close()
+      sys.exit(0)
+
+  def _get_variant(self, data_line):
+    try:
+      self._to_child.write(data_line + '\n')
+      self._to_child.flush()
+      return self._convert_to_variant(next(self._vcf_reader))
+    except (MemoryError, IOError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _convert_to_variant(
```

Comment:
I think one line is easier to read (put `self` and `record` in the same line).

Reply:
Done